### PR TITLE
Implement Bids Spec files

### PR DESCRIPTION
**Pull request by pvandyken**
This is the first out of two PRs for the resolution of #245. Here, we add define the spec file and add a `0.0.0` file to represent the bids function as it's been until now. The api does not change at all here, but the bids function has been refactored to use the spec.

I also added an extensive test suite for the bids function.

# The spec

The spec takes the following form (in yaml)

```yaml
- entity: entityLongName
  tag: entity # optional
  dir: true # optional
```

The order of entries determines the order entities will be placed in by the bids function. `datatype`, `suffix`, `extension`, and `prefix` are considered nonstandard entities and must not be specified in the config. Unrecognized entities are, by default, placed at the end of the path before the suffix. However, at most one entry in the spec may be specified as:

```yaml
- entity: "*"
```

This indicates a placeholder determining the insertion point of unrecognized entities.

Currently, no validation of the above specifications is enforced. The `bids()` function is hard-coded to use the `0.0.0` spec, which I already know is valid. In the future, when we support custom spec files, we will need to implement validation.

### CODE REVIEW

**Comment by pvandyken** [comment]
In fact, this PR does introduce a slight behaviour change. Previously, if `root` was specified as `.`, it would be dropped from the resulting path. So:

```py
bids(root=".", subject="001") == "sub-001/sub-001"
```

Now, the `.` will be kept as a path element:

```py
bids(root=".", subject="001") == "./sub-001/sub-001"
```

Practically speaking, this shouldn't make much of a difference, as the paths are still canonically the same (It got me in some of my tests, but they are more strict about paths being exactly correct). But I think this change is more correct than previous behaviour.

**Review by kaitj — COMMENTED** [review]
At a glance, this seems largely good to me. A few minor comments.

File: pyproject.toml [hunk 1]
```diff
@@ -93,7 +93,7 @@ build-backend = "poetry_dynamic_versioning.backend"
 [tool.poe.tasks]
 setup = "pre-commit install"
 quality = { shell = "isort snakebids && black snakebids && ruff snakebids && pyright" }
-fix = { shell = "isort snakebids && black snakebids && ruff --fix snakebids"}
+fix = { shell = "ruff --fix snakebids && isort snakebids && black snakebids"}
```
[the remark this answers is not in the document]
Comment (pvandyken) [reply pyproject.toml:96]: `ruff` sometimes affects the order of import sorts (e.g. removing unused imports), and may potentially unblack code, so this order ensures isort and black are both satisfied

**Review by tkkuehn — APPROVED** [review]
One minor comment and a question, but this looks good and I won't need to see it again.

File: snakebids/tests/strategies.py [hunk 2]
```diff
@@ -28,6 +28,10 @@
 valid_entities: tuple[str] = tuple(BidsConfig.load("bids").entities.keys())
 
 
+def nothing() -> Any:
```
[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/tests/strategies.py:31]: Strictly type checking, there was a problem with the hypothesis implementation so I wrapped it to save myself headaches

File: snakebids/__init__.py [hunk 3]
```diff
@@ -1,22 +1,25 @@
-__submodules__ = ["core"]
-
 __version__ = "0.0.0"
 
+__submodules__ = ["core", "paths"]
+__explicit__ = ["factory_specs"]
+
+from snakebids.paths import specs as factory_specs
```
[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/__init__.py:6]: Ahh, so `from snakebids.paths import bids` is handled by mkinit... Not isort compliant necessarily, but simpler to manage if I keep them separate

File: snakebids/paths/presets.py [hunk 4]
```diff
@@ -143,100 +178,41 @@ def bids(
...
-#this adds constraints to the bids naming
-wildcard_constraints:  **snakebids.get_wildcard_constraints(
-    config['pybids_inputs']
-)
+    tail = f"_{suffix}{extension or ''}" if suffix else extension or ""
 
-# ---- end snakebids boilerplate --------------------------------------------
-"""
-    )
+    return os.path.join(*path_parts) + tail
```
[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/paths/presets.py:218]: I don't know of a PurePath.join command. The alternative as I can see is `return str(Path(*path_parts)) + tail`. Do you find that preferable?

**Review by kaitj — APPROVED** [review]
Should be good - up to you about the minor comment but pending fixing the doc build error, this lgtm.

File: snakebids/paths/presets.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/paths/presets.py:218]: Yeah, `PurePath` is fine, it's the `.join`. There's a `.joinpath`, which is equivalent to the constructor method I showed above. And then any leading `.` is removed, so it changes behaviour.

**Pull request merged**